File: ch02/9斐波那契数列/fibonacci.cpp
class Solution {
public:
    //需要注意的是Fibonacci数列的第一项和第二项都都是1
    int Fibonacci(int n) {
        int FiboN = 0;
        int Fibo1 = 1;
        int Fibo2 = 1;
        if(n <= 0)
            return 0;
        if(1 == n or 2 == n)
            return 1;
        
        for(int i = 3; i <= n; i++) {
            FiboN = Fibo1 + Fibo2;
            Fibo1 = Fibo2;
            Fibo2 = FiboN;
        }
        
        return FiboN;
    }
};

File: ch02/9斐波那契数列/fibonacci.py
# -*- coding:utf-8 -*-
class Solution:
    # 用递推实现，不用递归的原因是递归有重复计算，存在很大的时间复杂度开销
    def Fibonacci(self, n):
        # write code here
        f1, f2, fn= 1, 1, 0
        if n == 0:
            return fn
        if n == 1 or n == 2:
            return f1
        
        for i in range(3, n + 1):
            fn = f1 + f2
            f1, f2 = f2, fn
        return fn